Add message of cancelled quest in party chat

HydeHunter2 opened the pull request:
Fixes #11093

### Changes
- A message of cancelling a quest that's still in the invitation phase will be displayed in the party chat

![github](https://user-images.githubusercontent.com/46248839/55645335-fa96fa00-57e0-11e9-8296-0feba814c40b.png)

- Change the comment due to function's renaming (`questCancel`->`cancelQuest`)

----
UUID: af1477b6-509a-4611-b851-d7a8e3a9c02a

HydeHunter2 committed:
Add message of cancelled quest in party chat

Issue #11093

HydeHunter2 committed:
Delete trailing spaces

For successful passing the test

HydeHunter2 commented:
While writing tests, I noticed that `partyMembers[1]` doesn't accept the invitation to quest:

https://github.com/HabitRPG/habitica/blob/f09b65e1084bacad76d954dac62424c58f86cbbf/test/api/v3/integration/quests/POST-groups_groupid_quests_cancel.test.js#L99-L110

Although in tests for aborting quest, `partyMembers[1]` accepts the invitation:

https://github.com/HabitRPG/habitica/blob/f09b65e1084bacad76d954dac62424c58f86cbbf/test/api/v3/integration/quests/POST-groups_groupid_quests_abort.test.js#L88-L91

Could I fix this in my commit, which adds tests?

Alys commented:
@HydeHunter2 In the 'cancels a quest' test, it's actually correct that `partyMembers[1]` hasn't accepted the invitation. :) The cancel command can be done only while the quest is still in the invitation stage, so in the test it's important to have a party member who hasn't accepted or rejected yet (after all members have accepted/rejected, the invitation phase ends automatically).

But thank you for asking about this! I guess it isn't clear at first and it does look like an inconsistency. I wonder if it's worth adding a brief comment under line 101 to explain that `partyMembers[1]` doesn't accept so that the quest stays in the invitation stage.

HydeHunter2 commented:
Ok. I understood.
Then I'll add a brief comment.

Thanks for the clarification!

HydeHunter2 committed:
Add test of cancelled quest's message

Also, added an explanation that partyMembers[1] hasn't accepted the invitation in the 'cancels a quest' test

HydeHunter2 committed:
Fix: import Group

Import Group to pass Lint syntax test

paglias commented on website/server/controllers/api-v3/quests.js:
```diff
@@ -363,12 +363,17 @@ api.cancelQuest = {
 
+    let questName = questScrolls[group.quest.key].text('en');
+    const newChatMessage = group.sendChat(`\`${user.profile.name} cancelled the party quest ${questName}.\``);
+    await newChatMessage.save();
```
thanks @HydeHunter2 ! Just one quick change is needed, could you move this save to the Promise.all down at line 380? So we don't block the other save operations while waiting for this one? Thanks

HydeHunter2 replied:
Of course

HydeHunter2 committed:
Move save function to Promise.all

HydeHunter2 commented:
Like this?
https://github.com/HabitRPG/habitica/blob/710aabbf4760233dd99ada44a6c2e16c0fe7857d/website/server/controllers/api-v3/quests.js#L373-L381

paglias commented on website/server/controllers/api-v3/quests.js:
```diff
@@ -372,12 +372,12 @@ api.cancelQuest = {
     group.markModified('quest');
 
     let [savedGroup] = await Promise.all([
+      newChatMessage.save(),
```
thanks! last thing, putting newChatMessage.save() at the beginning of the array will cause `savedGroup` to get its value instead of `group.save()`. It should be at least in second place inside the array

paglias commented on website/server/controllers/api-v3/quests.js:
```diff
@@ -372,12 +372,12 @@ api.cancelQuest = {
     group.markModified('quest');
 
     let [savedGroup] = await Promise.all([
+      newChatMessage.save(),
```
thanks! last thing, putting newChatMessage.save() at the beginning of the array will cause savedGroup to get its value instead of group.save(). It should be at least in second place inside the array

HydeHunter2 replied:
Sorry that I couldn't do it from the first commit(

Thanks for the review!

HydeHunter2 committed:
Fix moving save to Promise.all

paglias commented:
@HydeHunter2  thanks for this! I'm merging the PR and awarding your contributor tier.

Note that further tiers require increasing amounts of work from one to the next, so Tier 2 may require a couple of PRs or a larger PR to attain. But keep helping out and we'll express our gratitude accordingly!
